flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/GenericExactlyOnceSink.java

```diff
@@ -0,0 +1,173 @@
+public abstract class GenericExactlyOnceSink<IN extends Tuple> extends AbstractStreamOperator<IN> implements OneInputStreamOperator<IN, IN> {
+	private AbstractStateBackend.CheckpointStateOutputView out;
+	private TypeSerializer<IN> serializer;
+	protected TypeInformation<IN> typeInfo;
+
+	private ExactlyOnceState state = new ExactlyOnceState();
+
+	/**
+	 * Saves a handle in the state.
+	 * @param checkpointId
+	 * @throws IOException
+	 */
+	private void saveHandleInState(final long checkpointId) throws IOException {
+		//only add handle if a new OperatorState was created since the last snapshot/notify
+		if (out != null) {
+			out.writeByte(0); //EOF-byte
+			StateHandle<DataInputView> handle = out.closeAndGetHandle();
+			if (state.pendingHandles.containsKey(checkpointId)) {
+				state.pendingHandles.get(checkpointId).add(handle);
+			} else {
+				ArrayList<StateHandle<DataInputView>> list = new ArrayList<>();
+				list.add(handle);
+				state.pendingHandles.put(checkpointId, list);
+			}
+			out = null;
+		}
+	}
+
+	@Override
+	public StreamTaskState snapshotOperatorState(final long checkpointId, final long timestamp) throws Exception {
+		StreamTaskState taskState = super.snapshotOperatorState(checkpointId, timestamp);
+		saveHandleInState(checkpointId);
+		taskState.setFunctionState(state);
+		return taskState;
+	}
+
+	@Override
+	public void restoreState(StreamTaskState state, long recoveryTimestamp) throws Exception {
+		super.restoreState(state, recoveryTimestamp);
+		this.state = (ExactlyOnceState) state.getFunctionState();
+		out = null;
+	}
+
+	@Override
+	public void notifyOfCompletedCheckpoint(long checkpointId) throws Exception {
+		super.notifyOfCompletedCheckpoint(checkpointId);
+		saveHandleInState(checkpointId);
+
+		synchronized (state.pendingHandles) {
+			Set<Long> pastCheckpointIds = state.pendingHandles.keySet();
+			Set<Long> checkpointsToRemove = new HashSet<>();
+			for (Long pastCheckpointId : pastCheckpointIds) {
+				if (pastCheckpointId <= checkpointId) {
+					List<StateHandle<DataInputView>> handles = state.pendingHandles.get(pastCheckpointId);
+					for (StateHandle<DataInputView> handle : handles) {
+						DataInputView in = handle.getState(getUserCodeClassloader());
+						while (in.readByte() == 1) {
+							IN value = serializer.deserialize(in);
+							sendValue(value);
+						}
+						checkpointsToRemove.add(pastCheckpointId);
+					}
+				}
+			}
+			for (Long toRemove : checkpointsToRemove) {
+				state.pendingHandles.remove(toRemove);
+			}
+		}
+	}
+
+	/**
+	 * Write the given element into the backend.
+	 * @param value value to be written
+	 * @throws Exception
+     */
+	protected abstract void sendValue(IN value) throws Exception;
+
+	@Override
+	public void processElement(StreamRecord<IN> element) throws Exception {
+		IN value = element.getValue();
+		if (serializer == null) {
+			typeInfo = TypeExtractor.getForObject(value);
+			serializer = typeInfo.createSerializer(new ExecutionConfig());
```

Comment:
Of course you should automatically extract the `TypeInformation` and then create the corresponding `TypeSerializer` when you create the sink. But you shouldn't pass the `TypeInformation` into the sink and ship it to the cluster.